src/fromager/context.py

```diff
@@ -123,9 +124,39 @@ def add_to_build_order(
             # Requirement and Version instances that can't be
             # converted to JSON without help.
             json.dump(self._build_stack, f, indent=2, default=str)
-        constraints.write_from_build_order(
-            self._constraints_filename, self._build_stack
+        # constraints.write_from_build_order(
+        #     self._constraints_filename, self._build_stack
+        # )
+
+    def update_dependency_graph(
+        self,
+        parent_req: Requirement | None,
+        req_type: str,
+        req: Requirement,
+        version: Version,
+    ):
+        logger.debug(
+            "recording %s dependency %s -> %s %s",
+            req_type,
+            parent_req.name if parent_req else "(toplevel)",
+            req.name,
+            version,
         )
+        # If we have no parent requirement, we are processing an installation
+        # requirement even though it is called "toplevel" elsewhere. So, change
+        # the type. Also set the parent_ßkey to an empty string as a unique value
+        # that can't be mistaken for a real package.
+        if parent_req is None:
+            parent_key = ""
+            req_type = "install"
+        else:
+            parent_key = canonicalize_name(parent_req.name)
+        parent_edges = self.all_edges.setdefault(parent_key, [])
```

Comment:
should we move this to initialization and instead use `defaultdict(list)`?

Reply:
I wasn't sure how to express that in the type. Maybe it's still just a dict?

Reply:
yes I think it is still a dict

Reply:
It's a dict:
```python
>>> collections.defaultdict.__mro__
(<class 'collections.defaultdict'>, <class 'dict'>, <class 'object'>)
```

Reply:
Done